airbyte-integrations/connectors/source-exchange-rate-api/source_exchange_rate_api/source.py

```diff
@@ -0,0 +1,151 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+import datetime
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple, Union
+
+import requests
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream, IncrementalMixin
+from airbyte_cdk.sources.streams.http import HttpStream
+
+from source_exchange_rate_api.authenticator import HTTPApiKeyHeaderAuth
+
+
+class ExchangeRateApiStream(HttpStream, ABC):
+    url_base = "https://api.apilayer.com/exchangerates_data/"
+
+    def __init__(self, config: Mapping[str, Any], *args, **kwargs):
+        super(ExchangeRateApiStream, self).__init__(*args, **kwargs)
+        self.config: Mapping[str, Any] = config
+
+    @property
+    @abstractmethod
+    def primary_key(self) -> Union[str, List[str]]:
+        return super().primary_key
```

Comment:
an abstract method shouldn't really have an implementation

Reply:
Fixed, but python allows to have an actual implementation for abstract method. I would say this is an virtual method